Evaluate type before comparing equality of WordNet and other objects

vivekscl opened the pull request:
Fixes #1944 

As explained in the commit message,

When comparing the equality of WordNet objects with other objects, it is assumed that the other object is a WordNet object and hence the name attribute of the other object is compared with the name attribute of the WordNet object even though the other object might not be a WordNet object.

This will result in an AttributeError exception to be thrown in the case where the other object is not a WordNet object. For instance, when comparing a WordNet object with a string, since the string does not contain a name attribute, an exception is thrown.

To prevent this from happening, the type of the objects should be evaluated before checking for equality like so:
```python
def __eq__(self, other):
        return type(self) == type(other) and self._name == other._name

    def __ne__(self, other):
        return not self.__eq__(other)

    def __lt__(self, other):
        return type(self) == type(other) and self._name < other._name
```

iliakur commented:
imho, this goes against the duck-typing philosophy that Python embraces.

vivekscl commented:
>imho, this goes against the duck-typing philosophy that Python embraces.

Correct me if I'm wrong, but by duck-typing, you mean that we should not check the whether the object is of a certain type but whether the **object behaves like the object that has that specific attribute or method** right?  For instance, using the duck example, `don’t check whether it IS-a duck: check whether it QUACKS-like-a duck and WALKS-like-a duck`.

In this case, this code would violate duck-typing.
```python
type(self) == type(other)
```

So we should find a workaround such as:
```python
def __eq__(self, other):
        try:
            self._name == other._name
        except TypeError:
            # deal with error
```

iliakur commented:
Yup! Also bear in mind that `other._name` starts with an underscore, so it's not a "public" attribute, i.e. I wouldn't in the first place even rely on it being present. Are there attributes of the WordNet object that don't start with an underscore?

vivekscl commented:
>Are there attributes of the WordNet object that don't start with an underscore?

In the WordNet object, all fields start with an underscore, but there are methods that return these fields. So we should use them instead?

vivekscl committed:
Evaluate type before comparing equality of WordNet and other objects

When comparing the equality of WordNet objects with other objects, it 
is assumed that the other object is a WordNet object and hence the name 
attribute of the other object is compared with the name attribute of 
the WordNet object even though the other object might not be a WordNet 
object.

This will result in an AttributeError exception to be thrown in the 
case where the other object is not a WordNet object. For instance when 
comparing a WordNet object with a string, since the string does not 
contain a name attribute, an exception is thrown.

To prevent this from happening, the type of the objects should be 
evaluated before checking for equality.

vivekscl commented:
@stevenbird  I made new commits for this pull request. But since this was closed, may I know if there were any errors in the pull request? :smile:

Jay-Patankar commented:
@stevenbird can I take this issue if its still open.
